internal/kubernetes/drain/simulator.go

```
@@ -0,0 +1,171 @@
+	client     client.Client
+	podFilter  kubernetes.PodFilterFunc
+
+	nodeResultCache utils.TTLCache[simulationResult]
+	podResultCache  utils.TTLCache[simulationResult]
+}
+
+type simulationResult struct {
+	result bool
+	reason string
+}
+
+var _ DrainSimulator = &drainSimulatorImpl{}
+
+func NewDrainSimulator(
+	ctx context.Context,
+	client client.Client,
+	indexer *index.Indexer,
+	store kubernetes.RuntimeObjectStore,
+	podFilter kubernetes.PodFilterFunc,
+) DrainSimulator {
+	simulator := &drainSimulatorImpl{
+		store:      store,
+		podIndexer: indexer,
+		pdbIndexer: indexer,
+		client:     client,
+		podFilter:  podFilter,
+
+		// TODO i'm not very sure if both of the caches make sense, as they will cache the same result
+		nodeResultCache: utils.NewTTLCache[simulationResult](3*time.Minute, 10*time.Second),
+		podResultCache:  utils.NewTTLCache[simulationResult](3*time.Minute, 10*time.Second),
+	}
+
+	go simulator.nodeResultCache.StartCleanupLoop(ctx)
+	go simulator.podResultCache.StartCleanupLoop(ctx)
+
+	return simulator
+}
+
+func (sim *drainSimulatorImpl) SimulateDrain(ctx context.Context, node *corev1.Node) (bool, error) {
+	if result, exist := sim.nodeResultCache.Get(string(node.GetUID())); exist {
+		return result.result, errors.New(result.reason)
+	}
+
+	pods, err := sim.podIndexer.GetPodsByNode(ctx, node.GetName())
+	if err != nil {
+		return false, err
+	}
+
+	reasons := []string{}
+	for _, pod := range pods {
+		// TODO handle error
+		passes, _, _ := sim.podFilter(*pod)
+		if !passes {
+			continue
+		}
+
+		if result, exist := sim.podResultCache.Get(string(pod.GetUID())); exist {
+			if !result.result {
+				reasons = append(reasons, result.reason)
+			}
+			continue
+		}
+
+		if res, err := sim.SimulatePodDrain(ctx, pod); err != nil {
+			// TODO add suceeded/failed pod drain simulation count metric
+			reason := fmt.Sprintf("Cannot drain pod '%s', because: %v", pod.GetName(), err)
+			reasons = append(reasons, reason)
```

Instead of accumulating all the errors for all the pods, we could also let the simulation fail at this point.